# Mount drift and periodic error: what the logs say

Goal: a solid, evidence-based model of the Polaris's tracking drift before building PEC on it.

Pipeline (all in `utility/`, tested in `tests/test_pec_*`):

1. **Catalog** (`catalog_logs.py`) -- every archived log split into steady-tracking segments on one target.
2. **Segment datasets** (`extract_segments.py`) -- per segment: RA/Dec drift (the driver's `total_accum`,
   i.e. guide corrections + PEC applied, so PEC's own effect is removed) and the three motor angles.
3. **Hypothesis tests** (`pe_analysis.py`) -- does the periodic drift repeat with each motor's angle
   (a worm on that motor) or with time?
4. **Theta-space PEC** -- a worm profile (drift as a function of the motor angles) added to the driver's EMA,
   scored with the same benchmark as `analyse_pec_catalog.ipynb`: the drift left for the guider per
   correction interval.

Logs come from the session registry (`utility/sessions.toml`).

Re-run top to bottom after adding logs to `logs/archive/` and registering them (`uv run python sessions.py --add-new`).
Sections 1-2 rebuild the catalog and segment datasets (about 25 s) when the registry is newer than them, or with
`REBUILD = True`.

In [149]:
import os, sys, glob, re
import numpy as np
import pandas as pd
import plotly.express as px
import plotly.graph_objects as go
from plotly.subplots import make_subplots
sys.path.insert(0, os.path.abspath(os.path.join('.', '..', 'driver')))
sys.path.insert(0, os.path.abspath('.'))
import importlib, catalog_logs, extract_segments, pe_analysis, sessions
for m in (sessions, pe_analysis, catalog_logs, extract_segments):     # dependencies first: extract_segments imports pe_analysis
    importlib.reload(m)
from catalog_logs import catalog
from extract_segments import extract_all
from sessions import catalog_notes, catalog_phd2
from pe_analysis import (load_segments, theta_scan, time_scan, MOTORS, PecModel, WORM_THETA, worm_benchmark,
                         relative_to, worm_angle_control, fit_worm, motor_errors, partial_motor_errors,
                         sky_weights, worm_phase_table, phase_consistency, motor_direction, pass_turns,
                         auto_trend_degree, windowed_worm, fit_worm_given)
from extract_segments import DEFAULT_SITE

LOG_DIR = os.path.abspath(os.path.join('..', 'logs', 'archive'))
SEG_DIR = os.path.join(LOG_DIR, 'segments')
CATALOG_CSV = os.path.join(LOG_DIR, 'catalog_segments.csv')
PHD2_LOGS = catalog_phd2()   # session -> PHD2 guide log for the same night (from sessions.toml; used for PEC-off segments)
_INDEX = os.path.join(SEG_DIR, 'index.csv')
REBUILD = (not os.path.exists(_INDEX)                                       # or set True to force
           or os.path.getmtime(sessions.DEFAULT_REGISTRY) > os.path.getmtime(_INDEX))
MIN_MINUTES = 60          # segments used for the hypothesis tests
TEMPLATE = 'plotly_dark'

## 0. Where the driver's position comes from (firmware)

From `polestar_app` (decompiled sources in `Polaris_Internal_Polestar/`, call graph from the unstripped ELF):

- **517** = the MCU's motor angles as reported over the gimbal UART (MCU `'='` type 4: yaw total, pitch,
  extra-axis total), sent by `SP_PushGimbalPosToApp`.
- **518** = a fixed *box* attitude multiplied by **those same motor angles**, every cycle
  (`SensorMngTask` AHRS_CODE: `SP_BoxAhraUpdateByCode` -> `GetCameraAhrs` -> `SP_SendRotateVectorToApp`).
  `SP_BoxAhraUpdateByCode` is pure quaternion maths (rotate about Z by yaw, Y by pitch, multiply): no sensor reads.
- The box attitude is set once in BOX_INIT by `SP_BoxAhraInit` from the **accelerometer only** (tilt), and
  again only on calibration or compass alignment (527) -- a step change, not drift.
- The IMU fusion path (`SP_BoxAhraUpdate` -> Madgwick with gyro + accelerometer + magnetometer; Mahony;
  `SP_AhrsUpdate`) is **never called** -- its addresses appear only in debug tables.

- Single-star alignment (`SP_CalAUNoTilt`) and compass alignment are one-off maths on the box rotation, not fusion.
- AHRS data (`'K'`) reaches the MCU only from Benro's own goto/tracking tasks, which the driver doesn't run under
  advanced tracking. The MCU has **no IMU and no digital encoder**: it reads a sin/cos analogue rotor-angle
  sensor per motor (2 ADC channels each) and references no I2C/SPI peripheral.

**So the AHRS adds no drift while tracking**: the driver's theta values are the MCU's motor angles plus a
constant. Drift has to come from what the MCU's motor-shaft angle cannot see: the gear train after the
motor (worm periodic error, backlash, any gear-ratio mismatch) and the alignment model.

## 1. Catalog of archived logs

One row per steady-tracking segment (split at gotos, slews, jogs, rotates, parks and tracking changes;
first 60 s skipped). `usable` = at least 45 min with a pose source. `pose_source`/`theta_source` say where
the motor angles come from (KFLOG, SGLOG, PECLOG pose through the inverse kinematics, or the target rebuilt).

In [150]:
if REBUILD or not os.path.exists(CATALOG_CSV):
    cat = catalog(LOG_DIR, catalog_notes())     # notes from sessions.toml
    cat.to_csv(CATALOG_CSV, index=False)
cat = pd.read_csv(CATALOG_CSV, parse_dates=['start', 'end'])
use = cat[cat['usable']].copy()
print(f"{len(cat)} segments in {cat['session'].nunique()} sessions; {len(use)} usable, "
      f"{use['duration_min'].sum() / 60:.1f} h")
cols = ['session', 'segment', 'start', 'duration_min', 'driver_version', 'pec', 'n_sync_guide', 'sync_interval_s',
        'pulse_guiding', 'pose_source', 'alt_mean', 'roll_mean', 'm1_dps_hr', 'm2_dps_hr', 'm3_dps_hr', 'notes']
show = use[cols].copy()
show['session'] = show['session'].str.replace('alpaca.', '', regex=False)
show['start'] = show['start'].dt.strftime('%Y-%m-%d %H:%M')
show.round(1)

333 segments in 56 sessions; 60 usable, 119.7 h


,session,segment,start,duration_min,driver_version,pec,n_sync_guide,sync_interval_s,pulse_guiding,pose_source,alt_mean,roll_mean,m1_dps_hr,m2_dps_hr,m3_dps_hr,notes
6,greg_Beta3.1_08_02_comms_problems,6,2026-08-02 03:35,70.4,v2.2.0 Beta 3,on,16,216.4,False,peclog,42.6,-2.1,6.6,11.0,4.3,comms problems with lots of lag
16,jdm_Beta3.1_08_02_sg_peclog,9,2026-08-02 09:30,181.1,v2.2.0 Beta 3.1,on,219,36.1,False,peclog,75.5,38.5,52.1,5.0,3.8,sync guiding with logs(pec)
21,jdm_Beta3.1_08_02_sg_peclog,14,2026-08-02 12:44,396.4,v2.2.0 Beta 3.1,on,564,41.9,False,peclog,46.5,17.5,16.2,2.5,13.6,sync guiding with logs(pec)
26,jdm_Beta3.1_08_03_sg_peclog,4,2026-08-03 09:33,54.4,v2.2.0 Beta 3.1,on,79,40.0,False,peclog,69.5,39.5,9.4,11.9,4.0,sync guiding with logs(pec)
29,jdm_Beta3.1_08_03_sg_peclog,7,2026-08-03 11:01,76.9,v2.2.0 Beta 3.1,on,138,34.3,False,peclog,17.8,51.9,6.1,12.1,3.9,sync guiding with logs(pec)
32,jdm_Beta3.1_08_03_sg_peclog,10,2026-08-03 13:03,51.8,v2.2.0 Beta 3.1,on,86,38.1,False,peclog,38.6,21.1,21.3,2.5,17.8,sync guiding with logs(pec)
33,jdm_Beta3.1_08_03_sg_peclog,11,2026-08-03 13:55,51.0,v2.2.0 Beta 3.1,unknown,0,NaN,False,target,NaN,NaN,NaN,NaN,NaN,sync guiding with logs(pec)
34,jdm_Beta3.1_08_03_sg_peclog,12,2026-08-03 14:47,72.2,v2.2.0 Beta 3.1,unknown,0,NaN,False,target,NaN,NaN,NaN,NaN,NaN,sync guiding with logs(pec)
37,jdm_Beta3.1_08_03_sg_peclog,15,2026-08-03 16:10,188.9,v2.2.0 Beta 3.1,on,295,39.7,False,peclog,36.3,-22.4,16.4,10.4,10.4,sync guiding with logs(pec)
39,jdm_Beta4.3_08_30_kfposlog,1,2026-08-30 20:00,76.3,v2.2.0 Beta 4.3,on,69,79.7,False,kflog,31.9,-54.1,4.1,10.6,7.8,"tracking with logs(kf,pos) and socket force cl..."


### Pose coverage

Each motor's mean rate per segment. A worm on a motor gives a time period of THETA / rate, so segments
where **one** motor dominates, and the **same** motor at **different** rates, are what separate the hypotheses.

In [151]:
use['fastest'] = use[['m1_dps_hr', 'm2_dps_hr', 'm3_dps_hr']].fillna(-1).idxmax(axis=1).str[:2].str.upper()
use['n_fast'] = (use[['m1_dps_hr', 'm2_dps_hr', 'm3_dps_hr']] >= 8).sum(axis=1)
fig = make_subplots(rows=1, cols=3, subplot_titles=['M1 vs M3 rate', 'M2 vs M3 rate', 'Hours by number of fast motors (>= 8 deg/hr)'])
hover = use['session'].str.replace('alpaca.', '', regex=False) + ' #' + use['segment'].astype(str)
fig.add_trace(go.Scatter(x=use['m1_dps_hr'], y=use['m3_dps_hr'], mode='markers', text=hover,
                         marker=dict(size=np.sqrt(use['duration_min']) * 1.5, color=use['alt_mean'], colorscale='Viridis',
                                     colorbar=dict(title='alt', x=0.3)), name='segments'), row=1, col=1)
fig.add_trace(go.Scatter(x=use['m2_dps_hr'], y=use['m3_dps_hr'], mode='markers', text=hover,
                         marker=dict(size=np.sqrt(use['duration_min']) * 1.5, color=use['alt_mean'], colorscale='Viridis', showscale=False),
                         showlegend=False), row=1, col=2)
h = use.groupby('n_fast')['duration_min'].sum() / 60
fig.add_trace(go.Bar(x=h.index.astype(str), y=h.values, showlegend=False), row=1, col=3)
fig.update_xaxes(title_text='M1 deg/hr', row=1, col=1); fig.update_yaxes(title_text='M3 deg/hr', row=1, col=1)
fig.update_xaxes(title_text='M2 deg/hr', row=1, col=2); fig.update_xaxes(title_text='fast motors', row=1, col=3)
fig.update_layout(height=450, width=1400, template=TEMPLATE, title='Segment motor rates (marker size ~ duration, colour = altitude)')
fig.show()

## 2. Segment datasets

Per usable segment with a drift signal: drift (arcsec, the driver's correction convention, joined across
PEC model resets) and motor angles on the same time base. Drift sources: modern/legacy PECLOG
`total_accum`, SGLOG, PHD2 pulses (PEC-off segments with a guide log), or the SYNC GUIDING lines.

In [152]:
if REBUILD:
    extract_all(CATALOG_CSV, LOG_DIR, SEG_DIR, PHD2_LOGS)
idx = pd.read_csv(os.path.join(SEG_DIR, 'index.csv'))
print(f"{len(idx)} segment datasets")
print(idx.groupby(['drift_source', 'theta_source']).size().rename('segments').to_string())

54 segment datasets
drift_source  theta_source
peclog        kflog            7
              peclog          37
              sglog            6
phd2          target           1
sglog         sglog            2
sync_lines    target           1


Pick any segment to look at its drift and motor angles:

In [153]:
SEGMENT = 'alpaca.jdm_Beta7.1_09_29_pulseguide__seg4.csv'     # one of idx['file']
d = pd.read_csv(os.path.join(SEG_DIR, SEGMENT))
fig = make_subplots(rows=2, cols=1, shared_xaxes=True, vertical_spacing=0.08,
                    subplot_titles=['Drift (arcsec, slow trend removed)', 'Motor angles relative to start (deg)'])
t = d['t_sec'] / 60
for col, name in (('drift_ra_arcsec', 'RA'), ('drift_dec_arcsec', 'Dec')):
    y = d[col] - np.polyval(np.polyfit(d['t_sec'], d[col], 2), d['t_sec'])
    fig.add_trace(go.Scatter(x=t, y=y, name=f'{name} drift'), row=1, col=1)
for i in (1, 2, 3):
    fig.add_trace(go.Scatter(x=t, y=d[f'theta{i}'] - d[f'theta{i}'].iloc[0], name=f'M{i}'), row=2, col=1)
fig.update_xaxes(title_text='minutes', row=2, col=1)
fig.update_layout(height=650, width=1400, template=TEMPLATE, hovermode='x unified',
                  title=f"{SEGMENT} -- drift from {d['drift_source'].iloc[0]}, angles from {d['theta_source'].iloc[0]}")
fig.show()

## 3. Worm (motor angle) or time?

Both hypotheses scored on the same segments (>= `MIN_MINUTES`, duplicate captures dropped):

- **Angle scan**: for each motor and candidate worm angle THETA, the mean periodogram power (RA and Dec,
  slow trend removed) at the period that worm gives in each segment, THETA / that motor's rate.
- **Time scan**: the mean power at one fixed time period in every segment.

Background power is roughly the median of each curve. Caution from the tests (`test_pec_pe_analysis.py`):
a purely time-based signal can also produce an angle-scan peak (each segment hits *some* THETA = T x rate),
so the two scans have to be compared with each other, not read alone.

In [154]:
segs = load_segments(SEG_DIR, min_minutes=MIN_MINUTES)
print(f"{len(segs)} segments, {len({s.session for s in segs})} sessions")
THETAS = np.arange(3.0, 12.01, 0.25)
PERIODS = np.arange(10, 80.1, 1) * 60
angle = theta_scan(segs, THETAS)
times = time_scan(segs, PERIODS)
fig = make_subplots(rows=1, cols=2, subplot_titles=['Angle scan: power at THETA / motor rate', 'Time scan: power at a fixed period'])
for m, c in zip(MOTORS, ('royalblue', 'mediumseagreen', 'orange')):
    fig.add_trace(go.Scatter(x=angle.index, y=angle[m], name=m, line=dict(color=c)), row=1, col=1)
fig.add_vline(x=360 / 54, line=dict(color='gray', dash='dot'), row=1, col=1)
fig.add_trace(go.Scatter(x=times.index / 60, y=times['power'], name='time', line=dict(color='white')), row=1, col=2)
fig.update_xaxes(title_text='THETA, deg of motor rotation (dotted: 54-tooth worm, 6.67 deg)', row=1, col=1)
fig.update_xaxes(title_text='period, min', row=1, col=2)
fig.update_yaxes(title_text='mean power', row=1, col=1)
fig.update_layout(height=450, width=1400, template=TEMPLATE)
fig.show()
summary = pd.DataFrame({m: {'peak THETA': angle[m].idxmax(), 'peak power': angle[m].max(), 'median': angle[m].median()} for m in MOTORS}).T
summary.loc['time'] = {'peak THETA': times['power'].idxmax() / 60, 'peak power': times['power'].max(), 'median': times['power'].median()}
summary['peak / median'] = summary['peak power'] / summary['median']
summary.round(3)   # 'peak THETA' for the time row is the period in minutes

32 segments, 23 sessions


,peak THETA,peak power,median,peak / median
M1,6.50,0.204,0.105,1.941
M2,6.00,0.252,0.151,1.668
M3,5.75,0.260,0.098,2.652
time,74.00,0.216,0.122,1.773


### Per segment: power at each motor's worm period vs at 28 min

The decisive cases are segments where a worm period is far from 28 min. A worm on that motor shows power
at its own period; a time-based source shows power at 28 min whatever the motors do.

In [155]:
PEAK_THETA = 6.3          # deg: where the angle scans above peak
rows = []
for s in segs:
    pg = s.periodograms(); pmin, pmax = pg['range']; r = s.rates
    rec = {'segment': s.name.replace('alpaca.', ''), 'hours': round(s.hours, 1),
           'M1 deg/hr': round(r[0], 1), 'M2 deg/hr': round(r[1], 1), 'M3 deg/hr': round(r[2], 1)}
    for ax in ('ra', 'dec'):
        per, pw = pg[ax]
        P = lambda T: round(float(np.interp(T, per, pw)), 2) if pmin <= T <= pmax else np.nan
        rec[f'{ax} @28min'] = P(28 * 60)
        for i, m in enumerate(MOTORS):
            rec[f'{ax} @worm {m}'] = P(PEAK_THETA / r[i] * 3600) if r[i] >= 5 else np.nan
        rec[f'{ax} median'] = round(float(np.median(pw)), 2)
    rows.append(rec)
per_seg = pd.DataFrame(rows)
per_seg

,segment,hours,M1 deg/hr,M2 deg/hr,M3 deg/hr,ra @28min,ra @worm M1,ra @worm M2,ra @worm M3,ra median,dec @28min,dec @worm M1,dec @worm M2,dec @worm M3,dec median
0,jdm_Beta3.1_08_02_sg_peclog#9,2.5,52.3,4.5,3.2,0.01,0.00,NaN,NaN,0.03,0.00,0.00,NaN,NaN,0.04
1,jdm_Beta3.1_08_02_sg_peclog#14,6.6,15.5,2.9,13.1,0.03,0.08,NaN,0.05,0.01,0.00,0.01,NaN,0.00,0.01
2,jdm_Beta3.1_08_03_sg_peclog#7,1.2,6.5,12.5,3.8,0.13,NaN,0.10,NaN,0.08,0.36,NaN,0.38,NaN,0.06
3,jdm_Beta3.1_08_03_sg_peclog#15,3.1,16.3,9.8,10.1,0.10,0.02,0.19,0.29,0.03,0.02,0.00,0.02,0.03,0.05
4,jdm_Beta4.3_08_30_kfposlog#1,1.1,4.2,10.6,7.8,0.19,NaN,NaN,NaN,0.10,0.14,NaN,NaN,NaN,0.04
5,jdm_Beta4.3_08_30_kfposlog#4,1.1,1.3,11.2,8.8,0.04,NaN,NaN,NaN,0.08,0.38,NaN,NaN,NaN,0.10
6,jdm_Beta7.1_09_29_pulseguide#4,1.8,1.7,3.1,13.0,0.78,NaN,NaN,0.89,0.02,0.21,NaN,NaN,0.26,0.12
7,jdm_Beta7.1_09_29_pulseguide#10,3.4,7.6,0.9,15.8,0.02,0.01,NaN,0.01,0.02,0.01,0.09,NaN,0.02,0.02
8,mark_Beta3.1_08_06#2,1.1,19.5,4.6,14.7,0.13,0.25,NaN,0.09,0.10,0.31,0.50,NaN,0.25,0.24
9,mark_Beta4.0_08_18#1,1.2,29.9,3.9,24.1,0.01,0.29,NaN,0.10,0.06,0.05,0.60,NaN,0.02,0.07


## 4. Theta-space PEC, one more time -- in motor space

A worm error is an error in **one motor's** output angle, `e_i(theta_i)` = sin/cos of that motor's worm phase
(harmonics 1 and 2). The MCU can't see it: its angle comes from a sensor on the motor shaft, divided by 960
(design doc §5.5). On the sky it is a small rotation about that motor's axis; split along the celestial pole, the
Dec axis and the boresight (as the driver's equatorial axes), the first two parts are the RA and Dec corrections
the driver logs (the boresight part is field rotation, which no guider sees). So **RA/Dec drift =
`sky_weights(pose)` . `e(theta)`**, and the profile is fitted on RA and Dec jointly in motor space, where it holds
at any pose. PEC with a profile: the driver's EMA (tau 7.5 min) follows the drift with the profile removed, and the
profile adds its own rate from the current motor angles.

`WORM_THETA` = 6.0 deg: a 60-tooth wheel, as 960:1 = 16 x 60 (the 54-tooth / 6.67 deg guess came from a 972:1
estimate, ruled out in §5.5). The controls below test other tooth counts.

Scored like `analyse_pec_catalog.ipynb` (drift left for the guider per interval), relative to **EMA 7.5 min**,
with the profile fitted:

| Tier | Fitted on | Question |
|---|---|---|
| **hindsight** | the scored segment itself | upper bound -- not something PEC can do |
| **same session** | the session's other segments | would a learnt profile survive a goto? |
| **other sessions** | other nights of the same mount, raw MCU motor angles only | does the profile repeat night to night? |
| **live** | the segment so far (refit every 10 min after 30 min) | what the driver could actually do |

`theta_source` matters for the cross-night tier: KFLOG, PECLOG `theta_raw` (from 2026-10-02) and SGLOG carry the raw MCU angles (518: M1 includes the session's
compass / SPA heading, so M1's worm phase differs per session); PECLOG `theta_raw - zeta_offset` (`peclog_zeta`) is the
MCU's own motor angle (517), the same every session; the PECLOG pose run through
the inverse kinematics includes that session's alignment, fine within a session but not across them. Latitude
comes from the session log (a client's `SiteLatitude`), else the default site.

In [156]:
EMA = PecModel('EMA tau 7.5m', 'dema', 0, 450, k=0.0)       # driver default (plain EMA: same as PecAxis, faster)
RAW = set(idx.loc[idx['theta_source'].isin(['kflog', 'peclog_raw', 'peclog_zeta', 'sglog']), 'session'] + '#' + idx['segment'].astype(int).astype(str))
OWNER = lambda session: next((o for o in ('mark', 'greg', 'vyskocil') if o in session), 'jdm')   # whose mount
SAME_MOUNT = lambda a, b: OWNER(a) == OWNER(b)
_lat = {}
def lat_of(session):
    if session not in _lat:
        found = []
        for p in glob.glob(os.path.join(LOG_DIR, f'{session}*.log')):
            with open(p, errors='replace') as f:
                found += re.findall(r"'SiteLatitude': '(-?[\d.]+)'", f.read())
        _lat[session] = float(found[-1]) if found else DEFAULT_SITE[0]
    return _lat[session]

bench_w = worm_benchmark(segs, EMA, lat_of, raw_theta=RAW, same_mount=SAME_MOUNT)
rel_w = relative_to(bench_w, EMA.name)
print(f"{bench_w['seg'].nunique()} segments ({len(RAW & set(bench_w['seg']))} with raw motor angles). "
      f"Drift left for the guider, pooled; 'vs ref' = relative to {EMA.name}, n = segments scored:")
order = ['no_pec', EMA.name, 'worm: hindsight', 'worm: same session', 'worm: other sessions', 'worm: live']
rel_w.loc[order].round(3)

32 segments (11 with raw motor angles). Drift left for the guider, pooled; 'vs ref' = relative to EMA tau 7.5m, n = segments scored:


30                     120                    300  \
                         n vs no PEC vs ref     n vs no PEC vs ref     n   
no_pec                32.0     1.000  1.209  32.0     1.000  1.489  32.0   
EMA tau 7.5m          32.0     0.827  1.000  32.0     0.672  1.000  32.0   
worm: hindsight       32.0     0.802  0.970  32.0     0.653  0.972  32.0   
worm: same session    17.0     1.043  1.139  17.0     1.062  1.385  17.0   
worm: other sessions  10.0     0.963  1.000  10.0     0.749  0.977  10.0   
worm: live            32.0     0.826  0.999  32.0     0.661  0.985  32.0   

                                       
                     vs no PEC vs ref  
no_pec                   1.000  1.631  
EMA tau 7.5m             0.613  1.000  
worm: hindsight          0.625  1.019  
worm: same session       1.125  1.570  
worm: other sessions     0.692  0.999  
worm: live               0.619  1.010

### Controls: is the gain specific to a worm angle?

For each assumed worm angle -- the tooth counts 960:1 allows with an integer first stage, the old 54-tooth guess,
and arbitrary angles -- the hindsight fit (left) and the **cross-night** fit (right: profile learnt on the mount's
other nights, scored on a held-out night; raw motor angles only). A real worm shows as a dip at its angle and not
elsewhere. The cross-night control is the stronger test: extra parameters can't help there.

In [157]:
CANDIDATES = {f'{n} teeth': 360 / n for n in (30, 32, 40, 48, 60, 64, 80, 96)}
CANDIDATES.update({'54 teeth (972 guess)': 360 / 54, 'arbitrary 5.2': 5.2, 'arbitrary 6.9': 6.9,
                   'arbitrary 8.3': 8.3, 'arbitrary 10.3': 10.3})
raw_segs = [s for s in segs if s.name in RAW]
ctl_hind = worm_angle_control(segs, tuple(CANDIDATES.values()), EMA, lat_of)
rows = []
for label, W in CANDIDATES.items():
    b = worm_benchmark(raw_segs, EMA, lat_of, raw_theta=RAW, same_mount=SAME_MOUNT, guide_intervals_s=(120,), worm_theta=W)
    b = b[b['model'].isin(['no_pec', EMA.name, 'worm: other sessions'])]
    per = b.pivot_table(index=['seg', 'axis'], columns='model', values='rms')
    v = (per['worm: other sessions'] / per[EMA.name]).dropna()
    rows.append({'worm': label, 'angle deg': round(W, 3), 'hindsight / EMA': ctl_hind[W],
                 'cross-night / EMA (pooled)': relative_to(b, EMA.name).loc['worm: other sessions', (120, 'vs ref')],
                 'cross-night median': v.median(), 'cross-night better in %': (v < 1).mean() * 100})
controls = pd.DataFrame(rows).sort_values('angle deg')
fig = go.Figure()
for col, c in (('hindsight / EMA', 'orange'), ('cross-night / EMA (pooled)', 'mediumseagreen')):
    fig.add_trace(go.Scatter(x=controls['angle deg'], y=controls[col], mode='markers+lines', name=col,
                             text=controls['worm'], line=dict(color=c)))
fig.add_hline(y=1.0, line=dict(color='gray', dash='dot'))
fig.update_layout(height=420, width=1100, template=TEMPLATE, xaxis_title='assumed worm angle (deg)',
                  yaxis_title='drift left for the guider, relative to EMA (120 s)', title='Worm angle controls')
fig.show()
controls.round(3)

,worm,angle deg,hindsight / EMA,cross-night / EMA (pooled),cross-night median,cross-night better in %
7,96 teeth,3.750,1.014,1.005,1.006,25.0
6,80 teeth,4.500,1.012,1.004,1.001,35.0
9,arbitrary 5.2,5.200,1.002,1.019,1.010,30.0
5,64 teeth,5.625,0.969,1.042,1.006,45.0
4,60 teeth,6.000,0.972,0.977,0.970,70.0
8,54 teeth (972 guess),6.667,0.990,1.033,1.001,45.0
10,arbitrary 6.9,6.900,0.987,1.017,1.012,15.0
3,48 teeth,7.500,1.001,1.041,1.009,40.0
11,arbitrary 8.3,8.300,1.006,1.002,1.002,40.0
2,40 teeth,9.000,0.993,1.017,1.005,25.0


### Each motor by its own angle

Every segment, one per slider step (arrow keys step through them): each motor's angle error, plotted against that motor's angle (top) and folded onto one worm turn
(bottom). The drift gives two numbers (RA, Dec) for three motors, so a sample-by-sample split mixes them -- the
**leakage** table printed below says how much of each motor's true error lands in each estimate at this segment's
poses (M1 and M3 mix strongly unless theta2 is large). The dots are therefore **partial residuals**: the motor's
profile from a joint fit of *this* segment (orange; motors are separated by their different worm periods) plus what
that fit leaves unexplained. Green: the profile learnt from the mount's **other sessions** (all but this one), only for segments with raw motor
angles -- pose-derived angles include each session's alignment, so their worm phase can't be compared across sessions.
Colour = direction of travel (white: angle decreasing, blue: increasing) -- after a reversal (e.g. M2 near the
meridian) the gear teeth bear on the other flank, so the two need not line up; samples are shown faint where they can't
separate the motor's worm from slow drift: the motor turns too slowly for its worm to repeat faster than the trend can
bend (rate < 6 deg / (segment hours / trend degree)), or the same-direction pass is under 2 worm turns (e.g. M2's
short climb after reversing), or (with `MIN_OWN_SHARE`) its estimate is mostly another motor's. A motor too slow to fit
in the segment is taken from the other sessions' profile (orange = green for it), so its worm isn't pushed onto
another motor. M1's dots are the least reliable: RA/Dec
can't see field rotation, which on this mount is mostly a mix of M1 and M3, so a sample-by-sample split gives M1 only
~40% of its own error (7% near the zenith) and much of M3's leftover and the drift -- seen as M1 cycles alternating
high and low. The joint fit (lines) separates them by their different worm periods; `MIN_OWN_SHARE` fades mixed samples.


In [158]:
WORM = 6.0
MIN_OWN_SHARE = 0.0     # e.g. 0.6 fades samples where a motor's estimate is mostly other motors (that is most of M1)
START = 'alpaca.jdm_Beta7.1_09_29_pulseguide#10'          # slider starts here
MAX_POINTS = 400                                          # per plotted trace
SRC = dict(zip(idx['session'] + '#' + idx['segment'].astype(int).astype(str), idx['theta_source']))
grid = np.linspace(0, WORM, 200)
fig = make_subplots(rows=2, cols=3, vertical_spacing=0.14, horizontal_spacing=0.06,
                    subplot_titles=[f'M{m + 1} by motor angle' for m in range(3)] + [f'M{m + 1} folded on one {WORM:g} deg worm turn' for m in range(3)])
SUBTITLES = [a.to_plotly_json() for a in fig.layout.annotations]     # kept on every slider step
spans, titles = [], []
for s in sorted(segs, key=lambda x: x.name):
    lat = lat_of(s.session)
    others = [(o.t, o.ra, o.dec, o.theta, lat_of(o.session)) for o in raw_segs
              if o.session != s.session and SAME_MOUNT(o.session, s.session)] if s.name in RAW else []
    cross = fit_worm(others, worm_theta=WORM) if others else None
    # motors too slow to fit here: taken from the other sessions' profile (else their worm is pushed onto M1/M3)
    own = (fit_worm_given((s.t, s.ra, s.dec, s.theta, lat), cross, trend_degree='auto') if cross is not None
           else fit_worm([(s.t, s.ra, s.dec, s.theta, lat)], worm_theta=WORM, trend_degree='auto'))
    pe = partial_motor_errors(s.t, s.ra, s.dec, s.theta, lat, own, trend_degree='auto')
    # a motor's samples count only while its worm repeats faster than the trend can bend: rate >= WORM / (hours / degree)
    min_rate = WORM / (s.hours / auto_trend_degree(s.theta, WORM, t=s.t))
    direction = motor_direction(s.t, s.theta, min_rate=min_rate)     # +1 / -1 by direction of travel, 0 while too slow
    passes = pass_turns(s.t, s.theta, WORM, min_rate=min_rate)        # worm turns of each sample's same-direction pass
    sign_any = motor_direction(s.t, s.theta, min_rate=0.0)            # direction of travel for colouring, however slow
    W_ = sky_weights(s.theta, lat)
    own_share = np.diagonal(np.einsum('nji,njk->nik', W_, np.linalg.solve(W_ @ np.transpose(W_, (0, 2, 1)), W_)),
                            axis1=1, axis2=2)             # (n, 3): how much of each motor's own error its estimate keeps
    share = own_share.mean(axis=0)
    pick = np.unique(np.linspace(0, len(s.t) - 1, min(len(s.t), MAX_POINTS)).astype(int))
    first = s.name == START
    start = len(fig.data)
    in_legend = set()                                     # each label once per segment, on the first motor that has it
    for m in range(3):
        th, e, dirn, ptn = s.theta[pick, m], pe[pick, m], direction[pick, m], passes[pick, m]
        clear = own_share[pick, m] >= MIN_OWN_SHARE          # where this motor's error isn't mixed with another's
        sense = sign_any[pick, m]                             # direction of travel, however slow
        good = (dirn != 0) & (ptn >= 2) & clear               # fast enough, long enough pass, not mixed
        o_ = np.argsort(th)
        g3 = np.zeros((len(grid), 3)); g3[:, m] = grid
        for keep, colour, opacity, label in (((sense <= 0) & good, 'white', 0.6, 'angle decreasing'),
                                             ((sense > 0) & good, 'deepskyblue', 0.6, 'angle increasing'),
                                             ((sense <= 0) & ~good, 'white', 0.18, 'faint: too slow, short pass or mixed'),
                                             ((sense > 0) & ~good, 'deepskyblue', 0.18, 'faint: too slow, short pass or mixed')):
            k = keep
            show = bool(k.any()) and label not in in_legend
            if show:
                in_legend.add(label)
            for row, x in ((1, th[k]), (2, np.mod(th[k], WORM))):
                fig.add_trace(go.Scatter(x=x, y=e[k], mode='markers', marker=dict(size=3, color=colour, opacity=opacity),
                                         name=label, legendgroup=label, showlegend=show and row == 1, visible=first),
                              row=row, col=m + 1)
        for prof, name, c in ((own, 'fit of this segment', 'orange'), (cross, 'learnt from the other sessions', 'mediumseagreen')):
            if prof is None:
                continue
            fig.add_trace(go.Scatter(x=th[o_], y=prof.motor_error(s.theta[pick][o_])[:, m], mode='lines', line=dict(color=c, width=2),
                                     name=name, legendgroup=name, showlegend=m == 0, visible=first), row=1, col=m + 1)
            fig.add_trace(go.Scatter(x=grid, y=prof.motor_error(g3)[:, m], mode='lines', line=dict(color=c, width=2),
                                     legendgroup=name, showlegend=False, visible=first), row=2, col=m + 1)
    spans.append((start, len(fig.data) - start))
    turns = np.abs(np.diff(s.theta[::max(1, len(s.t) // 400)], axis=0)).sum(axis=0) / WORM     # path length
    unfitted = [f'M{m + 1}' for m in range(3) if turns[m] < 2]
    titles.append((s.name.replace('alpaca.', ''),
                   f"{s.name.replace('alpaca.', '')[:46]:46s}  {SRC.get(s.name, '?'):11s} angles  {s.hours:4.1f} h  lat {lat:5.1f}<br>"
                   f"worm turns M1/M2/M3 {turns[0]:5.1f} {turns[1]:5.1f} {turns[2]:5.1f}   "
                   f"own share {share[0]:4.2f} {share[1]:4.2f} {share[2]:4.2f}   "
                   f"not fitted (< 2 turns): {', '.join(unfitted) if unfitted else '-':10s}  "
                   f"green line (profile learnt from the other sessions): {'shown' if cross is not None else 'n/a, pose-derived angles':24s}"))
INFO = lambda text: dict(text=text, xref='paper', yref='paper', x=0, y=1.16, xanchor='left', yanchor='bottom',
                         align='left', showarrow=False, font=dict(family='monospace', size=12, color='#DDDDDD'))
AUTORANGE = {f'{a}{k}.autorange': True for a in ('xaxis', 'yaxis') for k in ('', 2, 3, 4, 5, 6)}
steps = []
for (start, count), (label, title) in zip(spans, titles):
    visible = [False] * len(fig.data)
    visible[start:start + count] = [True] * count
    steps.append(dict(method='update', label=label, args=[{'visible': visible}, {'annotations': SUBTITLES + [INFO(title)], **AUTORANGE}]))
active = next((k for k, s in enumerate(sorted(segs, key=lambda x: x.name)) if s.name == START), 0)
fig.update_layout(legend=dict(orientation='h', yanchor='bottom', y=1.06, xanchor='left', x=0), margin=dict(l=80, r=30, t=150, b=60), 
                  sliders=[dict(active=active, steps=steps, pad=dict(t=60), font=dict(size=9),
                                currentvalue=dict(prefix='segment: ', font=dict(color='#EEEEEE', size=12)))],
                  annotations=SUBTITLES + [INFO(titles[active][1])], height=850, width=1500, template=TEMPLATE)
for m in range(3):
    fig.update_xaxes(title_text=f'theta{m + 1} (deg)', row=1, col=m + 1)
    fig.update_xaxes(title_text='worm phase (deg)', row=2, col=m + 1)
fig.update_yaxes(title_text='motor angle error (arcsec)', col=1)
fig.show()

### Is each motor's worm steady through the night? (sliding windows)

**What it answers:** the folded charts above assume each motor's worm has the *same* size and timing all night. If it
doesn't (e.g. it depends on pose or load), the fold smears. This chart checks that directly.

**How:** the segment is cut into overlapping windows (`WINDOW_MIN` long, starting every `STEP_MIN`). In each window
the three motors' worms are fitted together, and each point is one motor's result for that window:
- **top: amplitude** -- how big that motor's 6 deg worm error is (arcsec of motor angle, 1st harmonic);
- **bottom: phase** -- where in its worm turn the error peaks (deg).

**How to read it:** flat lines = a steady worm, which a fixed profile (the driver's worm feed-forward) can correct. A
trend or jump = the worm really changes during the night. Dashed = the profile learnt from the mount's other sessions
(raw-angle segments). Hollow markers = M1 and M3 turn at about the same speed in that window, so their worms repeat at
nearly the same rate and the fit can trade one for the other (don't trust those points).

**Why a motor is often missing:** a motor is fitted in a window only if it turns at least 1.5 worm turns (9 deg) in it.
M2 usually turns slowly (a few deg/hr), so it appears only in some segments; segments where no motor qualifies are left
out of the slider.

In [159]:
WINDOW_MIN, STEP_MIN = 90, 15
COLOURS = {'M1': 'royalblue', 'M2': 'mediumseagreen', 'M3': 'orange'}
fig = make_subplots(rows=2, cols=1, shared_xaxes=True, vertical_spacing=0.08,
                    subplot_titles=['worm amplitude (arcsec of motor angle, 1st harmonic)', 'worm phase (deg)'])
SUBTITLES = [a.to_plotly_json() for a in fig.layout.annotations]     # kept on every slider step
WIN_AUTORANGE = {'xaxis.autorange': True, 'xaxis2.autorange': True, 'yaxis.autorange': True}   # this chart's axes only
spans, titles, shown = [], [], []
for s in sorted(segs, key=lambda x: x.name):
    lat = lat_of(s.session)
    w = windowed_worm(s, lat, window_s=WINDOW_MIN * 60, step_s=STEP_MIN * 60, worm_theta=WORM)
    if not w['fitted'].any():
        continue                                          # no motor turns enough in any window: nothing to show
    shown.append(s.name)
    others = [(o.t, o.ra, o.dec, o.theta, lat_of(o.session)) for o in raw_segs
              if o.session != s.session and SAME_MOUNT(o.session, s.session)] if s.name in RAW else []
    cross = fit_worm(others, worm_theta=WORM) if others else None
    first = s.name == START
    start = len(fig.data)
    for m, colour in COLOURS.items():
        g = w[(w['motor'] == m) & w['fitted']]
        if g.empty:
            continue
        hollow = ~g['m1_m3_separable'] & g['motor'].isin(['M1', 'M3'])          # M1 and M3 at about the same speed
        for row, col in ((1, 'amp'), (2, 'phase')):
            fig.add_trace(go.Scatter(x=g['t_min'], y=g[col], mode='lines+markers', line=dict(color=colour),
                                     marker=dict(symbol=np.where(hollow, 'circle-open', 'circle'), size=8),
                                     name=m, legendgroup=m, showlegend=row == 1, visible=first,
                                     text=[f'{x:.1f} turns, {r:.0f} deg/hr' for x, r in zip(g['turns'], g['rate'])]),
                          row=row, col=1)
        if cross is not None:
            i0 = (int(m[1]) - 1) * 2 * len(cross.harmonics)
            a, b = cross.coef[i0], cross.coef[i0 + 1]
            for row, val in ((1, np.hypot(a, b)), (2, np.degrees(np.arctan2(b, a)) % 360)):
                fig.add_trace(go.Scatter(x=[g['t_min'].min(), g['t_min'].max()], y=[val, val], mode='lines',
                                         line=dict(color=colour, dash='dash'), name=f'{m} other sessions', legendgroup=m,
                                         showlegend=False, visible=first), row=row, col=1)
    spans.append((start, len(fig.data) - start))
    titles.append((s.name.replace('alpaca.', ''), f"{s.name.replace('alpaca.', '')[:46]:46s}  {SRC.get(s.name, '?'):11s} angles  "
                   f"{s.hours:4.1f} h   windows of {WINDOW_MIN} min every {STEP_MIN} min"))
steps = []
for (start, count), (label, title) in zip(spans, titles):
    visible = [False] * len(fig.data)
    visible[start:start + count] = [True] * count
    steps.append(dict(method='update', label=label, args=[{'visible': visible}, {'annotations': SUBTITLES + [INFO(title)], **WIN_AUTORANGE}]))
active = shown.index(START) if START in shown else 0
for k, tr in enumerate(fig.data):                         # show the active segment's traces to start with
    tr.visible = spans[active][0] <= k < spans[active][0] + spans[active][1]
fig.update_layout(legend=dict(orientation='h', yanchor='bottom', y=1.06, xanchor='left', x=0), margin=dict(l=80, r=30, t=150, b=60), 
                  sliders=[dict(active=active, steps=steps, pad=dict(t=50), font=dict(size=9),
                                currentvalue=dict(prefix='segment: ', font=dict(color='#EEEEEE', size=12)))],
                  annotations=SUBTITLES + [INFO(titles[active][1])], height=750, width=1400, template=TEMPLATE)
fig.update_yaxes(range=[0, 360], row=2, col=1)
fig.update_xaxes(title_text='minutes from segment start', row=2, col=1)
fig.show()

### Does the worm phase repeat night to night?

A fixed profile only works if each motor's worm phase is the same every session, i.e. the raw MCU angle reference
survives power cycles. For every raw-angle segment where a motor turns at least 2 worm turns: the 1st-harmonic
amplitude and phase from a joint fit of that segment, and the phase difference between its two halves (noise check).
Per motor, the clustering of the phases across sessions: R = 1 identical, ~0 random; p = chance of that much
clustering from random phases. Control: the same at arbitrary worm angles, where phases should not cluster.

In [160]:
PHASE_ANGLES = (WORM_THETA, 5.2, 6.9, 8.3, 4.5)                  # 6.0 and arbitrary controls
for_mount = [s for s in raw_segs if OWNER(s.session) == 'jdm']
tables = {W: worm_phase_table(for_mount, lat_of, worm_theta=W) for W in PHASE_ANGLES}
ph = tables[WORM_THETA]
summary = pd.DataFrame({(f'{W:g} deg', m): phase_consistency(g.loc[g['amp'] >= 10, 'phase'])
                        for W, t in tables.items() for m, g in t.groupby('motor')}).T
print('phase clustering per motor (segments with amplitude >= 10"):')
display(summary.round(3))
fig = make_subplots(rows=1, cols=3, specs=[[{'type': 'polar'}] * 3], subplot_titles=[f'M{m}' for m in (1, 2, 3)])
for k, m in enumerate(('M1', 'M2', 'M3')):
    g = ph[ph['motor'] == m]
    fig.add_trace(go.Scatterpolar(theta=g['phase'], r=g['amp'], mode='markers+text', marker=dict(size=10),
                                  text=g['session'].str.replace('alpaca.', '').str[:22], textfont=dict(size=8),
                                  name=m, showlegend=False), row=1, col=k + 1)
fig.update_layout(height=480, width=1400, template=TEMPLATE,
                  title=f'{WORM_THETA:g} deg worm: 1st-harmonic phase (angle) and amplitude (arcsec, radius) per segment')
fig.show()
ph.assign(seg=ph['seg'].str.replace('alpaca.', '')).round(1)

phase clustering per motor (segments with amplitude >= 10"):


n     mean      R      p
6 deg   M1  6.0  297.520  0.858  0.012
        M2  5.0  108.549  0.986  0.008
        M3  6.0  271.385  0.954  0.004
5.2 deg M1  5.0  276.461  0.714  0.078
        M2  6.0  312.339  0.072  0.969
        M3  6.0   42.100  0.205  0.777
6.9 deg M1  2.0  343.052  0.955  0.161
        M2  5.0  263.281  0.080  0.968
        M3  6.0  113.007  0.423  0.343
8.3 deg M1  3.0   92.600  0.560  0.390
        M2  4.0  147.798  0.430  0.478
        M3  5.0  237.224  0.159  0.881
4.5 deg M1  4.0  198.319  0.448  0.448
        M2  5.0  109.062  0.100  0.951
        M3  3.0  325.661  0.456  0.537

,seg,session,motor,turns,amp,phase,halves_diff
0,jdm_Beta7.1_09_29_pulseguide#10,alpaca.jdm_Beta7.1_09_29_pulseguide,M1,3.8,19.9,302.8,NaN
1,jdm_Beta7.1_09_29_pulseguide#10,alpaca.jdm_Beta7.1_09_29_pulseguide,M3,8.1,93.6,258.0,-101.2
2,soak_Beta4.4_09_01_sg_sglog_Dec10#4,alpaca.soak_Beta4.4_09_01_sg_sglog_Dec10,M3,3.2,77.6,285.3,0.0
3,soak_Beta4.4_09_01_sg_sglog_Dec10#5,alpaca.soak_Beta4.4_09_01_sg_sglog_Dec10,M3,3.4,62.3,283.8,0.0
4,soak_Beta4.4_09_03_sgpec_sgkfposlog_ngc2070#2,alpaca.soak_Beta4.4_09_03_sgpec_sgkfposlog_ngc...,M1,7.6,33.7,235.0,-35.1
5,soak_Beta4.4_09_03_sgpec_sgkfposlog_ngc2070#2,alpaca.soak_Beta4.4_09_03_sgpec_sgkfposlog_ngc...,M3,6.2,63.5,281.2,-41.5
6,soak_Beta4.4_09_03_sgpec_sglog_NGC55#1,alpaca.soak_Beta4.4_09_03_sgpec_sglog_NGC55,M1,2.0,22.9,306.4,NaN
7,soak_Beta4.4_09_03_sgpec_sglog_NGC55#1,alpaca.soak_Beta4.4_09_03_sgpec_sglog_NGC55,M2,3.4,47.9,109.1,0.0
8,soak_Beta4.4_09_06_sg_sglog#0,alpaca.soak_Beta4.4_09_06_sg_sglog,M1,2.1,59.8,300.9,0.0
9,soak_Beta4.4_09_06_sg_sglog#0,alpaca.soak_Beta4.4_09_06_sg_sglog,M2,3.8,73.4,93.7,0.0


## 5. Findings (2026-10-02)

- **AHRS:** no live IMU input while tracking (section 0). Theta = MCU motor angles + a constant.
- **Worm signature (section 3, RA/Dec periodograms):** the angle scans peak near 5.5-6.75 deg for all three
  motors, and in segments where one motor dominates the worm period
  holds power where 28 min has none (e.g. Mark 09-04 M2, 07-29 M1). Not decisive on its own: the time scan
  also has peaks (~30 and ~74 min), and a time-based signal can produce angle peaks.
- **Data cleaning (2026-10-03):** most sync-guided segments open with a burst of plate solves seconds apart (to seed
  PEC). The solved positions stay on target, yet `total_accum` climbs up to ~140"/min: it is PEC's logged correction
  (`pec_accum`), which each sync re-anchors away before the mount carries it out. In the per-motor charts this was a
  steep streak, larger than any worm. Segment extraction now counts only the residuals for syncs under 10 s apart
  (`extract_segments.PHANTOM_GAP_S`), which lines the burst up with the drift after it (misfit 189" -> 77"); duplicate
  captures are loaded once and trends are fitted per minute rather than per sample. Results below use these segments.
- **M2 and M3 have a 6.0 deg worm (60 teeth, as 960:1 = 16 x 60) whose phase repeats night to night.** Phase check
  (raw angles, 2026-09-01..09-29): M3 6 segments, R 0.93 (p ~0.006); M2 5 segments, R 0.98 (p ~0.008); at arbitrary
  worm angles nothing clusters. Learnt profile (all these nights): M2 +/-64", M3 +/-73" of motor angle. Cross-night
  control: learnt on the mount's other nights, only 6.0 deg (and 12 deg, whose 2nd harmonic is 6.0) beats EMA
  (pooled 0.976, median 0.956, better in 70%); every other tooth count, the 54-tooth 6.67 deg and arbitrary angles
  score 1.00-1.04.
- **M1 also clusters once bursts are corrected** (R 0.89, 6 segments, 4 nights), but these nights share a site and setup:
  M1's 518 angle includes each session's compass / SPA heading, so this may not hold elsewhere. An M1 term adds little
  out of sample (below), so profiles leave it out until zeta-logged nights (517, no heading) confirm it.
- **What a profile from other nights removes** (leave one session out, 10 raw-angle segments, M2+M3): the drift the
  guider must correct (slow trend removed) drops by a median 16% and the drift accumulated over 10 minutes by 26%,
  better in 7 of 10 segments (best ~50%); M3 alone 10% / 9%, M1+M2+M3 15% / 26%. Learnt within a segment (live) or
  from the session's other segments (pose-derived angles) it does not beat EMA -- it has to be learnt across nights.
- **Implication:** a fixed M2/M3 worm feed-forward, learnt across nights on the raw MCU angles, is built
  (`Config.pec_worm_ff`, `utility/learn_worm.py`, kinematics.md §2.6.VI); the on-sky A/B test is next.
- Earlier (2026-10-01, rolling 10-min forecast vs a straight line): neither the 34+17 min time shape nor free
  per-motor worm terms forecast better than a line; the worm terms overfit with 2+ fast motors. Time-domain
  predictors (`analyse_pec_catalog.ipynb` and the 2026-10-02 tests of AR and adaptive sine predictors) also gain
  only in hindsight. **EMA tau 7.5 min stays the PEC model**, with an M3 worm feed-forward as the one
  candidate addition.
- **Why the archive struggles:** most drift is sync-guided `total_accum` (sampled at plate solves, with steps
  and solve noise, PEC running), segments are 1-2 h, only one night has a PHD2 guide log.

Open questions:
- Does the amplitude depend on pose (gravity load on M2/M3) or rotation direction? Why no clear M1 worm?
- What makes the non-periodic drift (alignment model -- QUEST leaves 20-100' non-rigid residuals, design doc §5.5 --
  or backlash at reversals; 960:1 itself is confirmed to ~0.5%)? It is most of what PEC is following.
- Would PEC-off, pulse-guided nights (PHD2 guide log, `log_position` on) show a worm the sync-guided archive hides?
  Only one night (2026-09-29) has a guide log; its PEC-off part is short.
